osg-project-waittime/calculate-waittime.py

```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python3
+
+from elasticsearch import Elasticsearch
+from elasticsearch_dsl import Search, A, Q
+import datetime
+import pandas as pd
+import dateutil.parser as parser
+import argparse
+
+GRACC = "https://gracc.opensciencegrid.org/q"
+
+HOUR = 3600
+
+# A mapping of usernames to projects
+usernameToProject = {}
+
+def getUsersPerDay(starttime: datetime.datetime, endtime: datetime.datetime):
+    """
+    Get a table of user's usage by day.  The table will have the format:
+        Users       day     day+1   day+2   ...     day+n
+        <user>      10      0       0       ...     31
+    
+    Where the usage is in CoreHours.
+    """
+
+    # 6 months back
+    starttime = starttime - datetime.timedelta(days=15)
+    es = Elasticsearch(
+        [GRACC],
+        timeout=300,
+        use_ssl=True,
+        verify_certs=True
+    )
+
+    MAXSZ = 2 ** 30
+    index = "gracc.osg.summary"
+    s = Search(using=es, index=index)
+    # Starttime and endtime are both datetime objects
+    s = s.query(
+        "bool",
+        filter=[
+            Q("range", EndTime={"gte": starttime, "lt": endtime})
+            & Q("term", ResourceType="Payload")
+            & Q("term", VOName="osg")
+            & (Q("term", ProbeName="condor:login04.osgconnect.net") | 
+               Q("term", ProbeName="condor:login05.osgconnect.net"))
+        ],
+    )
+
+    bkt = s.aggs
+    bkt = bkt.bucket("EndTime", A('date_histogram', field="EndTime", calendar_interval="1d"))
+    bkt = bkt.bucket("DN", A("terms", size=MAXSZ, field="DN"))
+    bkt = bkt.bucket("ProjectName", A("terms", size=MAXSZ, field="ProjectName"))
+    bkt.metric("CoreHours", 'sum', field="CoreHours", missing=0)
+
+    print(s.to_dict())
+    response = s.execute()
+
+    results_dict = {}
+    for bucket in response.aggregations["EndTime"]["buckets"]:
+        date = bucket["key"]
+        results_dict[date] = {}
+        for user in bucket['DN']['buckets']:
+            username = user['key']
+            for project in user['ProjectName']['buckets']:
+                if username not in usernameToProject:
+                    usernameToProject[username] = []
+                if project['key'] not in usernameToProject[username]:
+                    usernameToProject[username].append(project['key'])
+                results_dict[date][username] = project['CoreHours']['value']
+        #print(bucket.to_dict())
+
+    df = pd.DataFrame(results_dict)
+    df = df.fillna(0)
+    df.columns = pd.to_datetime([x*1000000 for x in df.columns])
```

Comment:
This looks curious - are the original `df.columns` values in seconds and the `pd.to_datetime()` function takes microseconds, or something?

Reply:
The [documented](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html) default is:
```
unitstr, default ‘ns’
```

But in the actual paragraph it says:
> The unit of the arg (D,s,ms,us,ns) denote the unit, which is an integer or float number. This will be based off the origin. Example, with unit=’ms’ and origin=’unix’ (the default), this would calculate the number of milliseconds to the unix epoch start.

So..., don't know, but this works I think.